static/app/utils/replays/hooks/useExtractedCrumbHtml.tsx

```diff
@@ -0,0 +1,137 @@
+import {useEffect, useState} from 'react';
+import first from 'lodash/first';
+import {Replayer} from 'rrweb';
+import {eventWithTime} from 'rrweb/typings/types';
+
+import type {Crumb} from 'sentry/types/breadcrumbs';
+import type ReplayReader from 'sentry/utils/replays/replayReader';
+
+// Copied from `node_modules/rrweb/typings/types.d.ts`
+enum EventType {
+  DomContentLoaded = 0,
+  Load = 1,
+  FullSnapshot = 2,
+  IncrementalSnapshot = 3,
+  Meta = 4,
+  Custom = 5,
+  Plugin = 6,
+}
+
+type Extraction = {
+  crumb: Crumb;
+  html: string;
+  timestamp: number;
+};
+
+type HookOpts = {
+  replay: ReplayReader;
+};
+function useExtractedCrumbHtml({replay}: HookOpts) {
+  const [breadcrumbRefs, setBreadcrumbReferences] = useState<Extraction[]>([]);
+
+  useEffect(() => {
+    const domRoot = document.createElement('div');
+    domRoot.className = 'sr-block';
+    const {style} = domRoot;
+    style.position = 'fixed';
+    style.inset = '0';
+    style.width = '0';
+    style.height = '0';
+    style.overflow = 'hidden';
+
+    document.body.appendChild(domRoot);
+
+    // Get a list of the breadcrumbs that relate directly to the DOM, for each
+    // crumb we will extract the referenced HTML.
+    const crumbs = replay
+      .getRawCrumbs()
+      .filter(crumb => crumb.data && 'nodeId' in crumb.data);
+
+    const rrwebEvents = replay.getRRWebEvents();
+
+    // Grab the last event, but skip the synthetic `replay-end` event that the
+    // ReplayerReader added. RRWeb will skip that event when it comes time to render
+    const lastEvent = rrwebEvents[rrwebEvents.length - 2];
+    const lastTimestamp = lastEvent.timestamp;
+
+    const isLastRRWebEvent = (event: eventWithTime) => lastTimestamp === event.timestamp;
+
+    const replayerRef = new Replayer(rrwebEvents, {
+      root: domRoot,
+      loadTimeout: 1,
+      showWarning: false,
+      blockClass: 'sr-block',
+      speed: 99999,
+      skipInactive: true,
+      triggerFocus: false,
+      plugins: [
+        new BreadcrumbReferencesPlugin({
+          crumbs,
+          isFinished: isLastRRWebEvent,
+          onFinish: rows => {
+            setBreadcrumbReferences(rows);
+            document.body.removeChild(domRoot);
```

Comment:
I suppose it's quite unlikely that the useEffect cleanup function gets called before the replay finishes?

Reply:
It's all async with timers afaik, so useEffect could get cleaned up before the replay finishes yeah. 

The issue we could see is the `setBreadcrumbReferences` call is made, but the react component is unmounted so the state change can't cause the re-render. That's solvable with a flag i guess.